From 8c73bc4be64f26dddde62dc864095ac09988d67b Mon Sep 17 00:00:00 2001
From: Stephen Herr <stephenherr@microsoft.com>
Date: Wed, 30 Nov 2022 15:25:43 +0000
Subject: [PATCH] Merged PR 7224968: Better control of signing services in apt
 repos

This PR:
1. Adds the ability to set default signing services to apt repos so we don't have to do our custom `pulp_label` hack and the pmcserver anymore.
1. Adds the ability to set an override signing service on a per-release basis to support `azurecore` and `azurecore-test`.
1. Teaches the pmcserver api to use it.

1 and 2 are changes in `pulp_deb` itself. There will be a necessary 1-time migration once this is released to move signing service information over to the new fields and away from our hack. Furthermore, because `Releases` are just Content types that are in immutable RepoVersions and therefore don't have an update api, it will be necessary to set their `default_signing_service` in the database for the 5 releases that need overrides.

Related work items: #16306468
---
 .../0021_add_default_signing_service.py       | 49 +++++++++++++++++++
 pulp_deb/app/models/__init__.py               |  4 +-
 pulp_deb/app/models/repository.py             |  6 +++
 pulp_deb/app/models/structure_content.py      | 16 +++++-
 .../app/serializers/content_serializers.py    | 12 +++++
 .../app/serializers/repository_serializers.py | 15 +++++-
 pulp_deb/app/tasks/publishing.py              |  8 ++-
 7 files changed, 102 insertions(+), 8 deletions(-)
 create mode 100644 pulp/container-assets/pulp_deb/pulp_deb/app/migrations/0021_add_default_signing_service.py

diff --git a/pulp_deb/app/migrations/0021_add_default_signing_service.py b/pulp_deb/app/migrations/0021_add_default_signing_service.py
new file mode 100644
index 0000000..ee21540
--- /dev/null
+++ b/pulp_deb/app/migrations/0021_add_default_signing_service.py
@@ -0,0 +1,49 @@
+# Generated by Django 3.2.16 on 2022-11-28 16:56
+
+from django.db import migrations, models
+import django.db.models.deletion
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("deb", "0020_add_release_fields"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="aptrepository",
+            name="signing_service",
+            field=models.ForeignKey(
+                null=True,
+                on_delete=django.db.models.deletion.PROTECT,
+                related_name="deb_aptrepository",
+                to="deb.aptreleasesigningservice",
+            ),
+        ),
+        migrations.AddField(
+            model_name="release",
+            name="signing_service",
+            field=models.ForeignKey(
+                null=True,
+                on_delete=django.db.models.deletion.PROTECT,
+                related_name="deb_release",
+                to="deb.aptreleasesigningservice",
+            ),
+        ),
+        migrations.AlterUniqueTogether(
+            name="release",
+            unique_together={
+                (
+                    "codename",
+                    "suite",
+                    "distribution",
+                    "version",
+                    "origin",
+                    "label",
+                    "description",
+                    "signing_service",
+                )
+            },
+        ),
+    ]
diff --git a/pulp_deb/app/models/__init__.py b/pulp_deb/app/models/__init__.py
index b54d677..1779585 100644
--- a/pulp_deb/app/models/__init__.py
+++ b/pulp_deb/app/models/__init__.py
@@ -7,6 +7,8 @@ from .content import (
     Package,
 )
 
+from .signing_service import AptReleaseSigningService
+
 from .structure_content import (
     Release,
     ReleaseArchitecture,
@@ -21,5 +23,3 @@ from .publication import AptDistribution, AptPublication, VerbatimPublication
 from .remote import AptRemote
 
 from .repository import AptRepository
-
-from .signing_service import AptReleaseSigningService
diff --git a/pulp_deb/app/models/repository.py b/pulp_deb/app/models/repository.py
index 1e25504..390ef35 100644
--- a/pulp_deb/app/models/repository.py
+++ b/pulp_deb/app/models/repository.py
@@ -1,8 +1,10 @@
+from django.db import models
 from pulpcore.plugin.models import Repository
 
 from pulpcore.plugin.repo_version_utils import remove_duplicates, validate_version_paths
 
 from pulp_deb.app.models import (
+    AptReleaseSigningService,
     AptRemote,
     GenericContent,
     InstallerFileIndex,
@@ -39,6 +41,10 @@ class AptRepository(Repository):
         AptRemote,
     ]
 
+    signing_service = models.ForeignKey(
+        AptReleaseSigningService, on_delete=models.PROTECT, null=True
+    )
+
     class Meta:
         default_related_name = "%(app_label)s_%(model_name)s"
 
diff --git a/pulp_deb/app/models/structure_content.py b/pulp_deb/app/models/structure_content.py
index 9849e09..7ea9125 100644
--- a/pulp_deb/app/models/structure_content.py
+++ b/pulp_deb/app/models/structure_content.py
@@ -4,7 +4,7 @@ from django.db import models
 
 from pulpcore.plugin.models import Content
 
-from pulp_deb.app.models import Package
+from pulp_deb.app.models import AptReleaseSigningService, Package
 
 
 BOOL_CHOICES = [(True, "yes"), (False, "no")]
@@ -26,11 +26,23 @@ class Release(Content):
     origin = models.TextField(null=True)
     label = models.TextField(null=True)
     description = models.TextField(null=True)
+    signing_service = models.ForeignKey(
+        AptReleaseSigningService, on_delete=models.PROTECT, null=True
+    )
 
     class Meta:
         default_related_name = "%(app_label)s_%(model_name)s"
         unique_together = (
-            ("codename", "suite", "distribution", "version", "origin", "label", "description"),
+            (
+                "codename",
+                "suite",
+                "distribution",
+                "version",
+                "origin",
+                "label",
+                "description",
+                "signing_service",
+            ),
         )
 
 
diff --git a/pulp_deb/app/serializers/content_serializers.py b/pulp_deb/app/serializers/content_serializers.py
index 520d0ce..ce06c6b 100644
--- a/pulp_deb/app/serializers/content_serializers.py
+++ b/pulp_deb/app/serializers/content_serializers.py
@@ -10,12 +10,14 @@ from pulpcore.plugin.serializers import (
     ContentChecksumSerializer,
     MultipleArtifactContentSerializer,
     NoArtifactContentSerializer,
+    RelatedField,
     SingleArtifactContentSerializer,
     SingleArtifactContentUploadSerializer,
     DetailRelatedField,
 )
 
 from pulp_deb.app.models import (
+    AptReleaseSigningService,
     BasePackage,
     GenericContent,
     InstallerFileIndex,
@@ -644,6 +646,15 @@ class ReleaseSerializer(NoArtifactContentSerializer):
     origin = CharField(required=False, allow_null=True, default=None)
     label = CharField(required=False, allow_null=True, default=None)
     description = CharField(required=False, allow_null=True, default=None)
+    signing_service = RelatedField(
+        help_text="A reference to an associated signing service. Overrides "
+        "AptRepository.signing_service but not AptPublication.signing_service.",
+        view_name="signing-services-detail",
+        queryset=AptReleaseSigningService.objects.all(),
+        many=False,
+        required=False,
+        allow_null=True,
+    )
 
     class Meta(NoArtifactContentSerializer.Meta):
         model = Release
@@ -655,6 +666,7 @@ class ReleaseSerializer(NoArtifactContentSerializer):
             "origin",
             "label",
             "description",
+            "signing_service",
         )
 
 
diff --git a/pulp_deb/app/serializers/repository_serializers.py b/pulp_deb/app/serializers/repository_serializers.py
index ab9211f..088a6bb 100644
--- a/pulp_deb/app/serializers/repository_serializers.py
+++ b/pulp_deb/app/serializers/repository_serializers.py
@@ -1,11 +1,12 @@
 from gettext import gettext as _
 from pulpcore.plugin.serializers import (
+    RelatedField,
     RepositorySerializer,
     RepositorySyncURLSerializer,
     validate_unknown_fields,
 )
 
-from pulp_deb.app.models import AptRepository
+from pulp_deb.app.models import AptReleaseSigningService, AptRepository
 
 from jsonschema import Draft7Validator
 from rest_framework import serializers
@@ -17,8 +18,18 @@ class AptRepositorySerializer(RepositorySerializer):
     A Serializer for AptRepository.
     """
 
+    signing_service = RelatedField(
+        help_text="A reference to an associated signing service. Used if "
+        "AptPublication.signing_service is not set",
+        view_name="signing-services-detail",
+        queryset=AptReleaseSigningService.objects.all(),
+        many=False,
+        required=False,
+        allow_null=True,
+    )
+
     class Meta:
-        fields = RepositorySerializer.Meta.fields
+        fields = RepositorySerializer.Meta.fields + ("signing_service",)
         model = AptRepository
 
 
diff --git a/pulp_deb/app/tasks/publishing.py b/pulp_deb/app/tasks/publishing.py
index 9130c8a..0948643 100644
--- a/pulp_deb/app/tasks/publishing.py
+++ b/pulp_deb/app/tasks/publishing.py
@@ -21,6 +21,7 @@ from pulpcore.plugin.models import (
 
 from pulp_deb.app.models import (
     AptPublication,
+    AptRepository,
     Package,
     PackageReleaseComponent,
     Release,
@@ -101,7 +102,7 @@ def publish(repository_version_pk, simple=False, structured=False, signing_servi
             publication.simple = simple
             publication.structured = structured
             publication.signing_service = signing_service
-            repository = repo_version.repository
+            repository = AptRepository.objects.get(pk=repo_version.repository.pk)
 
             if simple:
                 codename = "default"
@@ -126,6 +127,7 @@ def publish(repository_version_pk, simple=False, structured=False, signing_servi
                     description=repository.description,
                     label=repository.name,
                     version=str(repo_version.number),
+                    signing_service=repository.signing_service,
                 )
 
                 for package in Package.objects.filter(
@@ -168,6 +170,7 @@ def publish(repository_version_pk, simple=False, structured=False, signing_servi
                         version=release.version or str(repo_version.number),
                         suite=release.suite,
                         origin=release.origin,
+                        signing_service=(release.signing_service or repository.signing_service),
                     )
 
                     for prc in PackageReleaseComponent.objects.filter(
@@ -257,6 +260,7 @@ class _ReleaseHelper:
         description=None,
         suite=None,
         origin=None,
+        signing_service=None,
     ):
         self.publication = publication
         self.distribution = distribution
@@ -290,7 +294,7 @@ class _ReleaseHelper:
 
         self.architectures = architectures
         self.components = {component: _ComponentHelper(self, component) for component in components}
-        self.signing_service = publication.signing_service
+        self.signing_service = publication.signing_service or signing_service
 
     def add_metadata(self, metadata):
         artifact = metadata._artifacts.get()
-- 
2.31.1

